dkulib/parallelizer/parallelizer.py

```diff
@@ -0,0 +1,278 @@
+
+    It takes the `function` as input and:
+    - ensures it has a `batch` parameter which is a list of dict
+    - parses the response to extract results and errors using the `batch_response_parser` function
+    - handles errors from the function with two methods:
+        * (default) log the error message as a warning and return the row with error keys
+        * fail if there is an error
+
+    """
+    output_batch = deepcopy(batch)
+    if error_handling == ErrorHandling.FAIL:
+        response = function(batch=batch, **function_kwargs)
+        output_batch = batch_response_parser(batch=batch, response=response, column_names=column_names)
+        errors = [row[column_names.error_message] for row in batch if row[column_names.error_message] != ""]
+        if len(errors) != 0:
+            raise BatchError(f"Batch function {function.__name__} failed on: {batch} because of error: {errors}")
+    else:
+        try:
+            response = function(batch=batch, **function_kwargs)
+            output_batch = batch_response_parser(batch=batch, response=response, column_names=column_names)
+        except exceptions as error:
+            logging.warning(f"Batch function {function.__name__} failed on: {batch} because of error: {error}")
+            error_type = str(type(error).__qualname__)
+            module = inspect.getmodule(error)
+            if module is not None:
+                error_type = str(module.__name__) + "." + error_type
+            for row in output_batch:
+                row[column_names.response] = ""
+                row[column_names.error_message] = str(error)
+                row[column_names.error_type] = error_type
+                row[column_names.error_raw] = str(error.args)
+    return output_batch
+
+
+def convert_results_to_df(
+    input_df: pd.DataFrame,
+    results: List[Dict],
+    column_names: NamedTuple,
+    error_handling: ErrorHandling = ErrorHandling.LOG,
+    verbose: bool = DEFAULT_VERBOSE,
+) -> pd.DataFrame:
+    """Combine results (list of dict) with input dataframe
+
+    Helper function to the `parallelizer` main function
+
+    """
+    if error_handling == ErrorHandling.FAIL:
+        columns_to_exclude = [column_name for key, column_name in column_names._asdict().items() if "error" in key]
+    else:
+        columns_to_exclude = []
+        if not verbose:
+            columns_to_exclude = [column_names.error_raw]
+    output_schema = {**{column_name: str for column_name in column_names}, **dict(input_df.dtypes)}
+    output_schema = {
+        column_name: schema_type
+        for column_name, schema_type in output_schema.items()
+        if column_name not in columns_to_exclude
+    }
+    record_list = [
+        {column_name: results.get(column_name) for column_name in output_schema.keys()} for results in results
+    ]
+    column_list = [column_name for column_name in column_names if column_name not in columns_to_exclude]
```

Comment:
Not sure it would work here, as the order of columns matters. Does convert a list to a set and subtracting sets keep the original order or the first list?

Reply:
With ef2a78d I don't need to do this anymore